### PR TITLE
Add/dsl guide

### PR DESCRIPTION
**Pull request by cmontella**

### CODE REVIEW

File: guides/dsl.md [hunk 1]
```diff
@@ -0,0 +1,241 @@
+# Eve JavaScript/TypeScript DSL
+
+The forthcoming v0.3 release of Eve supports a domain specific langauge (DSL) that interacts with the Eve server, allowing you to intermix Javascript and Eve code. There are several benefits to using the DSL:
```
Comment (ibdknox) [inline guides/dsl.md]: there's no server at the moment, so probably worth mentioning this is client-side only for now

File: guides/dsl.md [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+});
+```
+
+for now you can find a list of functions in [src/runtime/stdlib.ts](https://github.com/witheve/Eve/blob/refactor-editor/src/runtime/stdlib.ts)
+
+The interface for wrapping functions for use within Eve is also revamped for the new runtime. When writing a function wrapper, you must ensure that the function meets the following qualifications
+
+1. The function must be referentially transparent, meaning given the same input, the function should return the same output.
+
+2. The function must be monotonic, meaning that additional inputs should never decrease the number of outputs.
```
Comment (ibdknox) [inline guides/dsl.md]: we don't really need to talk about functions being monotone here since referential transparency prevents that from being a problem. Our internal operations in Eve have to be mostly monotonic, but that doesn't extend to anyone else really.

File: guides/dsl.md [hunk 3]
```diff
@@ -0,0 +1,241 @@
+# Eve JavaScript/TypeScript DSL
+
+The forthcoming v0.3 release of Eve supports a domain specific langauge (DSL) that interacts with the Eve server, allowing you to intermix Javascript and Eve code. There are several benefits to using the DSL:
+
+1) The syntax is native Javascript, so if you know JS you can write in the DSL
+2) You can use the DSL with Javascript functions and libraries
```
Comment (ibdknox) [inline guides/dsl.md]: sort of, you can't use functions/libraries in the blocks but you can go the other way around if you wanted

File: guides/dsl.md [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+  return [
+    // bind a single record tagged "guest-list", with every friend as a guest
+    record("guest-list").add({guest: friend})
+  ];
+});
+```
+
+### remove()
+
+Removes attribute/value pairs from a record. This is equivalent to the Eve `-=` oeprator. For example:
```
Comment (ibdknox) [inline guides/dsl.md]: remove should only be used in commits

File: guides/dsl.md [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+- Remove an attribute `person.remove("salary")` <-> `person.salary := none`
+- Remove a record `person.remove()` <-> `person := none`
+
+## Subblocks
+
+In the DSL `not()`, `union()`, and `choose()` are subblocks, which have their own body. Let's look at each of these.
+
+### not()
+
+The `not()` sub block works similary to `not()` from the Eve syntax; it performs an anti-join on the records inside and outside of the `not()`. To use not, include it in the function list at the beginning of the block.
```
Comment (ibdknox) [inline guides/dsl.md]: param list

File: guides/dsl.md [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+    return [Math.sin(a/180 * Math.PI)];
+  }
+});
+```
+
+## Importing and Exporting Records 
+
+### Getting Data Into Eve - `inputEavs()`
+
+You can import raw EAVs into Eve with the `inputEavs()` function. Due to limitations of the host language, in order for it to work correctly, `inputEavs()` can only be used after you define all blocks.
```
Comment (ibdknox) [inline guides/dsl.md]: we can fix that

File: guides/dsl.md [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+
+`inputEavs()` takes as input a list of entity, attribute, value triples. The entity value identifies the record to which the attribute and value belong, so it must be unique to that record. For example:
+
+```javascript
+prog.inputEavs([[0,"tag","person"],[0,"name","Archibald"])
```
Comment (ibdknox) [inline guides/dsl.md]: @joshuafcole as part of your getting started stuff, it's probably worth standardizing on some version of just importing JS objects directly.

File: guides/dsl.md [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+import {Program} from "./runtime/dsl2";
+```
+
+Then you instantiate a new Eve program, naming it however you want.
+
+```javascript
+let program = new Program("program name");
+```
+
+Writing DSL code is similar to writing native Eve code; you attach blocks to the `program`, which search for records and then return records. Blocks in the DSL operate as a `bind` would in native Eve. Let's look at the DSL block syntax:
```
Comment (ibdknox) [inline guides/dsl.md]: @cmontella @joshuafcole maybe we should just rename block to bind?

**Review by ibdknox — DISMISSED** [review]

File: guides/dsl.md (hunk 8 above)
Comment (cmontella) [reply guides/dsl.md]: yeah, Josh and I had actually talked about this one. I think it's a good move.

File: guides/dsl.md [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+search
+  person = [#person]
+
+bind
+  [#greeting message: "Hello {{person.name}}"]
+```
+
+The identity of a `record()` is established by everything within the parentheses. Therefore, the record `record("greeting", {message: `Hello ${person.name}`})` will create one record for every `person.name`.
+
+If would like to commit records instead of binding them, you can use the `commit()` method in place of `block()`. The implicatino here is that you can no longer create blocks which both commit and bind records at the same time, which we've found is not a common case.
```
Comment (joshuafcole) [inline guides/dsl.md]: "implicatino"

File: guides/dsl.md [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+
+The DSL supports methods which mirror a couple of the Eve update operators. They are:
+
+### add()
+
+ dds attribute/value pairs to a record. These added values do not contribute to the identity of a record, allowing you to add multiple values toa single record. This is equivalent to the Eve `+=` operator. For example:
```
Comment (joshuafcole) [inline guides/dsl.md]: "dds"

Comment (joshuafcole) [reply guides/dsl.md]: "toa"

Comment (joshuafcole) [reply guides/dsl.md]: You can also use `.add()` on an entity variable, e.g. `friend.add("quality", "swell")`

File: guides/dsl.md [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+  let teacher = find("teacher", {name: "Smith"});
+  return [
+    artemis.remove("teacher").add("teacher", teacher)
+  ];
+});
+```
+
+In summary:
+
+- Add a value: `person.salary += 10` <-> `person.add("salary", 10)`
```
Comment (joshuafcole) [inline guides/dsl.md]: This one is backwards compared to the others.

File: guides/dsl.md [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+```
+
+### Functions
+
+The standard library in Eve has been redone in the DSL. To use library functions, you must now bring in "lib" explicitly when defining your block. From lib you can access the various standard library functions supported by the runtime so far. 
+
+```javascript
+program.block("display the student's full names", ({find, record, lib}) => {
+  find("angle" degrees)
+  let result = math.sin(degrees)
```
Comment (joshuafcole) [inline guides/dsl.md]: as imported, this would be `lib.math.sin(...)`. To use it as `math.sin, instead use this destructuring in the parameter list: `{find, record, lib:{math}}`.

File: guides/dsl.md (hunk 6 above)
Comment (joshuafcole) [reply guides/dsl.md]: Due to another fixable limitation you need to call inputEavs at least once to kick the program off.

File: guides/dsl.md (hunk 7 above)
Comment (joshuafcole) [reply guides/dsl.md]: Righto. The only issue is that it gets a bit clunky for graph-y things. In practice that probably doesn't come up often though?

File: guides/dsl.md (hunk 11 above)
Comment (joshuafcole) [reply guides/dsl.md]: I think this section should just be a translation guide from the existing syntax to the JS DSL. We should add find and record here, along with the statements and expressions and aggregates.

File: guides/dsl.md [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+    let student = find("student");
+    // lookup attributes and values related to each student
+    let {attribute, value} = lookup(student);
+    return [
+      // Add these attributes and values to the student, which creates a diff to which we can react
+      student.add(attribute, value)
+    ];
+  })
+  // React to each addition or removal
+  .asDiffs((diff) => {
```
Comment (joshuafcole) [inline guides/dsl.md]: Need a section on `asObjects` as well, since people will probably want to use that one more often than asDiffs

File: guides/dsl.md [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+    for(let [e, a, v] of diff.adds) {
+      // ... do something ...
+    }
+  });
+```
+
```
Comment (joshuafcole) [inline guides/dsl.md]: Missing a section on aggregates

**Review by joshuafcole — CHANGES_REQUESTED** [review]

File: guides/dsl.md (hunk 8 above)
Comment (joshuafcole) [reply guides/dsl.md]: done in https://github.com/witheve/Eve/pull/816, the guide is correct again. :)

File: guides/dsl.md (hunk 6 above)
Comment (joshuafcole) [reply guides/dsl.md]: `inputEavs` is now `inputEAVs` https://github.com/witheve/Eve/pull/815

File: guides/dsl.md (hunk 1 above)
Comment (joshuafcole) [reply guides/dsl.md]: Yeah let's change "Eve server" to "Eve runtime" here to not give people the wrong idea. We do have a server, but it's a dumb server. It just manages getting files from crazy places into your browser.

File: guides/dsl.md [hunk 15]
```diff
@@ -0,0 +1,289 @@
...
+2) You can interact with Javascript functions and libraries
+3) The DSL can be used "a la carte", so you can use as much of Eve as you need to for your project. Thus you can write an entire application in Eve, just use it as a datastore, or anywhere between. This makes it easy to integrate Eve with an existing Javascript application.
+4) You can easily import data into Eve. If you can get your data in Javascript, you can use it in Eve.
+
+## Using the DSL
+
+To use the DSL, import it into your application.
+
+```javascript
+import {Program} from "./runtime/dsl2";
```
Comment (joshuafcole) [inline guides/dsl.md]: People will be consuming from the `witheve` module, so the import is `import {Program} from "witheve"`.

File: guides/dsl.md [hunk 16]
```diff
@@ -0,0 +1,289 @@
...
+search
+  person = [#person]
+
+bind
+  [#greeting message: "Hello {{person.name}}"]
+```
+
+The identity of a `record()` is established by everything within the parentheses. Therefore, the record `record("greeting", {message: `Hello ${person.name}`})` will create one record for every `person.name`.
+
+If would like to commit records instead of binding them, you can use the `commit()` method in place of `bind()`. The implication here is that you can no longer create blocks which both commit and bind records at the same time.
```
Comment (joshuafcole) [inline guides/dsl.md]: Let's make a note somewhere that this guide is for people already comfortable with the Eve semantics, which have not changed. If a brand new user were to wander in and see `bind`, `commit`, `record` et al thrown around without explanation it'd be pretty confusing. Similarly, it'll help to ground existing users from the start by telling them that we work the exact same and just look different.

File: guides/dsl.md [hunk 17]
```diff
@@ -0,0 +1,289 @@
...
+});
+```
+
+You can also use `remove()` to completely delete records or attributes on a record. This is equivalent to `:= none` in Eve. For example:
+
+```javascript
+   // completely remove the target record
+   record.remove() 
+   // remove the attribute on the target record
+   record.remove("attribute")
```
Comment (joshuafcole) [inline guides/dsl.md]: Let's use an attribute other than "attribute" here to make sure it's obvious that this is an example value and not a magic keyword.

File: guides/dsl.md [hunk 18]
```diff
@@ -0,0 +1,289 @@
...
+
+ Adds attribute/value pairs to a record. These added values do not contribute to the identity of a record, allowing you to add multiple values to a single record. This is equivalent to the Eve `+=` operator. For example:
+
+```javascript
+program.bind("Invite classmates to my party", ({find, record}) => {
+  // search for records tagged "friend"
+  let student = find("student");
+  return [
+    // bind a single record tagged "guest-list", with every student as a guest
+    record("guest-list").add({guest: student})
```
Comment (joshuafcole) [inline guides/dsl.md]: Not sure if we need to make a point of it since this is basically always equally or more concise as you've written it, but we support both `.add("guest", student).add("foo", bar)` and `.add({guest: student, foo: bar})`

File: guides/dsl.md [hunk 19]
```diff
@@ -0,0 +1,289 @@
...
+```
+
+In summary:
+
+- Add a value: `person.add("salary", 10)` <-> `person.salary += 10`
+- Remove a value: `person.remove("salary, 10)` <-> `person.salary -= 10`
+- Set a value: `person.remove("salary").add("salary", 10)` <-> `person.salary := 10`
+- Remove an attribute `person.remove("salary")` <-> `person.salary := none`
+- Remove a record `person.remove()` <-> `person := none`
+
```
Comment (joshuafcole) [inline guides/dsl.md]: Feels a bit weird to have this summary be in the middle of the guide. I'd probably put it at the top as a TL;DR or the bottom as a summary. Either way I'd have one line for *everything*, including expressions, aggregates, chooses, and unions.

Comment (joshuafcole) [reply guides/dsl.md]: People will come in and (hopefully!) read the whole thing once, but this one liner section will likely be what they'll come back and refer to as they work with Eve.

File: guides/dsl.md [hunk 20]
```diff
@@ -0,0 +1,289 @@
...
+## Sub-Blocks
+
+In the DSL `not()`, `union()`, and `choose()` are sub-blocks, which have their own body. Let's look at each of these.
+
+### not()
+
+The `not()` sub block works similarly to `not()` from the Eve syntax; it performs an anti-join on the records inside and outside of the `not()`. To use not, include it in the function list at the beginning of the block.
```
Comment (joshuafcole) [inline guides/dsl.md]: this isn't the function list, it's the parameter list

File: guides/dsl.md [hunk 21]
```diff
@@ -0,0 +1,289 @@
...
+    });
+    return [
+        students.add("tag","good-standing")
+    ];
+})
+```
+
+### choose()
+
+Choose() and union() expressions are behind the mechanics of the if expression in the Eve syntax. In the DSL, we expose these directly. First, `choose()` takes a body of branches, each one comprised of a condition and a return value. The value of the choose corresponds to the first matching branch.
```
Comment (joshuafcole) [inline guides/dsl.md]: This is a bit confusingly worded. I'd describe it more like "choose takes a list of subblocks which contain any valid eve code to join, filter, mash, or compute their results. Each is executed in order, and the first to succeed is taken as the choose's return value. If no subblock succeeds, the row fails."

File: guides/dsl.md [hunk 22]
```diff
@@ -0,0 +1,289 @@
...
+    )
+    return [
+        student.add("letter-grade", {grade}),
+    ];
+});
+```
+
+### union()
+
+By contrast `union()` takes a body of branches and combines their output into a single variable. One use of union might be to normalize records from different data sources.
```
Comment (joshuafcole) [inline guides/dsl.md]: I'd reword this one to avoid body and branches like above.

File: guides/dsl.md [hunk 23]
```diff
@@ -0,0 +1,289 @@
...
+
+You can import raw EAVs into Eve with the `inputEAVs()` function. Currently, `inputEAVs()` mut be called at least once to initialize your program, and it can only be used after you define all blocks. These limitations will be lifted in future versions.
+
+`inputEAVs()` takes as input a list of entity, attribute, value triples. The entity value identifies the record to which the attribute and value belong, so it must be unique to that record. For example:
+
+```javascript
+prog.inputEavs([[0,"tag","person"],[0,"name","Archibald"])
+```
+
+This will create a record tagged "person" with the name attribute "Archibald".
```
Comment (joshuafcole) [inline guides/dsl.md]: We also have a convenience function that you can import from witheve called `appendAsEAVs`. It destructures an object into EAVs with a unique id for you.

```
import {appendAsEAVs} from "witheve";
let inputs = [];
appendAsEAVS(inputs, {tag: "person", name: "Chris", pets: ["Fifi", "Amumu"]});
// ...
prog.inputEAVs(inputs);
```

File: guides/dsl.md [hunk 24]
```diff
@@ -0,0 +1,289 @@
...
+
+### Getting Records out of Eve - Watchers
+
+Watchers are a new feature native to the DSL that allow you to monitor changes in specific records and react to them with a callback function. Like blocks, watchers also attach to a `program`.
```
Comment (joshuafcole) [inline guides/dsl.md]: Watchers aren't specific to the DSL -- They just integrate with whatever method you use for writing Eve blocks. In the future we may support the Eve syntax again or other more exotic ways of creating the watcher's blocks as well. :)

File: guides/dsl.md [hunk 25]
```diff
@@ -0,0 +1,289 @@
...
+  program.watch("Export student GPA", ({find, lookup, record}) => {
+    // search for records tagged student
+    let student = find("student" {name, GPA});
+    return [
+      // Add these attributes and values to the student, which creates a diff to which we can react
+      record("grade", {name, GPA})
+    ];
+  })
+  // Handle adds and removes as objects
+  .asObjects<{name:string, GPA:RawValue}>(({adds, removes}) => {
```
Comment (joshuafcole) [inline guides/dsl.md]: The type casty bit between `<...>` is typescript specific. If using raw JS, just leave that bit out and everything else remains the same. It's essentially just instructions to TypeScript telling it what you expect the Diffs to look like. I'd add a note in a comment right above it mentioning that.

File: guides/dsl.md [hunk 26]
```diff
@@ -0,0 +1,289 @@
...
+
+The forthcoming v0.3 release of Eve supports a domain specific langauge (DSL) that interacts with the Eve server, allowing you to intermix Javascript and Eve code. There are several benefits to using the DSL:
+
+1) The syntax is native Javascript, so if you know JS you can write in the DSL
+2) You can interact with Javascript functions and libraries
+3) The DSL can be used "a la carte", so you can use as much of Eve as you need to for your project. Thus you can write an entire application in Eve, just use it as a datastore, or anywhere between. This makes it easy to integrate Eve with an existing Javascript application.
+4) You can easily import data into Eve. If you can get your data in Javascript, you can use it in Eve.
+
+## Using the DSL
+
```
Comment (joshuafcole) [inline guides/dsl.md]: We need a little preface here saying that the guide is written using ES6. Specifically, we use and recommend TypeScript ^2.1.0 (which transpiles ES6 to cross-platform compatible javascript. Specifically, for concision and readability we make frequent use of destructuring [1] and arrow functions [2].

[1] https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment
[2] https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Functions#Arrow_functions

File: guides/dsl.md (hunk 1 above)
Comment (joshuafcole) [reply guides/dsl.md]: still need this one.

File: guides/dsl.md [hunk 27]
```diff
@@ -0,0 +1,314 @@
...
+- Set a value: `person.remove("salary").add("salary", 10)` <-> `person.salary := 10`
+- Remove an attribute `person.remove("salary")` <-> `person.salary := none`
+- Remove a record `person.remove()` <-> `person := none`
+- functions `lib.math.sin()` <-> `sin[]`
+- aggregates `gather().pers().count()` <-> `count[given per]` 
+
+## Using the DSL
+
+
+The Eve Javascript DSL is requires ES6. For added readability, we make frequent use of [destructuring][1] and [arrow functions][2] available in TypeScript 2.1.0 or later.
```
Comment (joshuafcole) [inline guides/dsl.md]: It doesn't require ES6, you can write it normally in ES5, we've just chosen to write in ES6 and thus the guide is written with ES6.

File: guides/dsl.md [hunk 28]
```diff
@@ -0,0 +1,314 @@
...
+3) The DSL can be used "a la carte", so you can use as much of Eve as you need to for your project. Thus you can write an entire application in Eve, just use it as a datastore, or anywhere between. This makes it easy to integrate Eve with an existing Javascript application.
+4) You can easily import data into Eve. If you can get your data in Javascript, you can use it in Eve.
+
+This DSL guide is for users who are already familiar with Eve semantics. For those new to Eve, we'll have more fundamental guides released in the near future.
+
+## Summary of DSL 
+
+- find a record: `find("person", {salary})` <-> `[#person salary]`
+- bind/commit a record: `record("person", {salary})` <-> `[#person salary]`
+- not: `not()` <-> `not()`
```
Comment (joshuafcole) [inline guides/dsl.md]: I'd have a very simple example filled in for each of these like you have above and below, e.g. `not(() => employee.fired)` <-> `not(employee.fired)`

File: guides/dsl.md [hunk 29]
```diff
@@ -0,0 +1,314 @@
...
+- choose: `choose()` <-> `if-then / else if-then`
+- union: `union()` <-> `if-then / if-then`
+- Add a value: `person.add("salary", 10)` <-> `person.salary += 10`
+- Remove a value: `person.remove("salary, 10)` <-> `person.salary -= 10`
+- Set a value: `person.remove("salary").add("salary", 10)` <-> `person.salary := 10`
+- Remove an attribute `person.remove("salary")` <-> `person.salary := none`
+- Remove a record `person.remove()` <-> `person := none`
+- functions `lib.math.sin()` <-> `sin[]`
+- aggregates `gather().pers().count()` <-> `count[given per]`
```
Comment (joshuafcole) [inline guides/dsl.md]: `pers` -> `per`

File: guides/dsl.md [hunk 30]
```diff
@@ -0,0 +1,314 @@
...
+- bind/commit a record: `record("person", {salary})` <-> `[#person salary]`
+- not: `not()` <-> `not()`
+- choose: `choose()` <-> `if-then / else if-then`
+- union: `union()` <-> `if-then / if-then`
+- Add a value: `person.add("salary", 10)` <-> `person.salary += 10`
+- Remove a value: `person.remove("salary, 10)` <-> `person.salary -= 10`
+- Set a value: `person.remove("salary").add("salary", 10)` <-> `person.salary := 10`
+- Remove an attribute `person.remove("salary")` <-> `person.salary := none`
+- Remove a record `person.remove()` <-> `person := none`
+- functions `lib.math.sin()` <-> `sin[]`
```
Comment (joshuafcole) [inline guides/dsl.md]: Would fill the argument in here for clarity (since lib functions are no longer keyword based.

File: guides/dsl.md (hunk 29 above)
Comment (joshuafcole) [reply guides/dsl.md]: Would fill in a full example here

File: guides/dsl.md [hunk 31]
```diff
@@ -0,0 +1,314 @@
...
+    });
+    return [
+        students.add("tag","good-standing")
+    ];
+})
+```
+
+### choose()
+
+Choose() and union() expressions are behind the mechanics of the if expression in the Eve syntax. In the DSL, we expose these directly. First, `choose()` takes a list of sub-blocks, which contain any valid Eve code to join, filter, or compute their results. Each sub-block executed in order until one is found valid. This return value of the first valid sub-block is taken as the return value.
```
Comment (joshuafcole) [inline guides/dsl.md]: "sub-block executed" -> "sub-block is executed"

File: guides/dsl.md [hunk 32]
```diff
@@ -0,0 +1,314 @@
...
+  program.watch("Export student GPA", ({find, lookup, record}) => {
+    // search for records tagged student
+    let student = find("student" {name, GPA});
+    return [
+      // Add these attributes and values to the student, which creates a diff to which we can react
+      record("grade", {name, GPA})
+    ];
+  })
+  // Handle adds and removes as objects
+  .asObjects(({adds, removes}) => {
```
Comment (joshuafcole) [inline guides/dsl.md]: The type bit you had here before is needed if you're writing in TypeScript. I'd just include it and add a note for users who aren't using TypeScript about what it is and that they should omit it.

File: guides/dsl.md [hunk 33]
```diff
@@ -0,0 +1,314 @@
...
+4) You can easily import data into Eve. If you can get your data in Javascript, you can use it in Eve.
+
+This DSL guide is for users who are already familiar with Eve semantics. For those new to Eve, we'll have more fundamental guides released in the near future.
+
+## Summary of DSL 
+
+- find a record: `find("person", {salary})` <-> `[#person salary]`
+- bind/commit a record: `record("person", {salary})` <-> `[#person salary]`
+- not: `not(() => person.salary)` <-> `not(person.salary)`
+- choose: `choose()` <-> `if-then / else if-then`
```
Comment (joshuafcole) [inline guides/dsl.md]: would have a tiny example for choose and union too

File: guides/dsl.md [hunk 34]
```diff
@@ -0,0 +1,314 @@
...
+- Set a value: `person.remove("salary").add("salary", 10)` <-> `person.salary := 10`
+- Remove an attribute `person.remove("salary")` <-> `person.salary := none`
+- Remove a record `person.remove()` <-> `person := none`
+- functions `lib.math.sin(number)` <-> `sin[degrees: number]`
+- aggregates `gather(person).per(person.dept).count()` <-> `count[given: person, per: person.dept]` 
+
+## Using the DSL
+
+
+This guide is written for ES6. For added readability, we make frequent use of [destructuring][1] and [arrow functions][2] available in TypeScript 2.1.0 or later.
```
Comment (joshuafcole) [inline guides/dsl.md]: The destructuring and arrow functions are part of ES6 (basically the next-gen of javascript). Typescript accepts ES6 syntax and translates it back into current-gen javascript, but there are a lot of other transpilers that users could choose to use instead. We recommend but don't require typescript.

File: guides/dsl.md [hunk 35]
```diff
@@ -0,0 +1,314 @@
...
+    let student = find("student" {name, GPA});
+    return [
+      // Add these attributes and values to the student, which creates a diff to which we can react
+      record("grade", {name, GPA})
+    ];
+  })
+  // Handle adds and removes as objects
+  .asObjects(({adds, removes}) => {
+    for(let key in adds) {
+      // ... do something ...
```
Comment (joshuafcole) [inline guides/dsl.md]: You should show that the shape of the diff matches the shape of the record created by the watch block here.
E.g.
```
let {name, GPA} = adds[id];
// ... do something ...
```

File: guides/dsl.md [hunk 36]
```diff
@@ -0,0 +1,314 @@
...
+    // search for records tagged student
+    let student = find("student" {name, GPA});
+    return [
+      // Add these attributes and values to the student, which creates a diff to which we can react
+      record("grade", {name, GPA})
+    ];
+  })
+  // Handle adds and removes as objects
+  .asObjects(({adds, removes}) => {
+    for(let key in adds) {
```
Comment (joshuafcole) [inline guides/dsl.md]: I'd rename `key` to `id` for clarity -- this key will be the id of the exported record. In many cases you don't care, but in some that's useful to know.

File: guides/dsl.md [hunk 37]
```diff
@@ -0,0 +1,316 @@
...
+
+This DSL guide is for users who are already familiar with Eve semantics. For those new to Eve, we'll have more fundamental guides released in the near future.
+
+## Summary of DSL 
+
+- find a record: `find("person", {salary})` <-> `[#person salary]`
+- bind/commit a record: `record("person", {salary})` <-> `[#person salary]`
+- not: `not(() => person.salary)` <-> `not(person.salary)`
+- choose: `choose(() => { person.salary; return 1; }, () => 0)` <-> `if person.salary then 1 else 0`
+- union: `union(() => { person.salary; return salary; }, () => person.wage; return wage)` <-> `if person.salary then salary if person.wage then wage`
```
Comment (joshuafcole) [inline guides/dsl.md]: For the DSL case you can simplify this to:

`union(() => person.salary, () => person.wage)` :)

File: guides/dsl.md (hunk 36 above)
Comment (joshuafcole) [inline guides/dsl.md]: I'd rename key to id for clarity -- this key will be the id of the exported record. In many cases you don't care, but in some that's useful to know.

File: guides/dsl.md [hunk 38]
```diff
@@ -0,0 +1,316 @@
...
+    // search for records tagged student
+    let student = find("student" {name, GPA});
+    return [
+      // Add these attributes and values to the student, which creates a diff to which we can react
+      record("grade", {name, GPA})
+    ];
+  })
+  // Handle adds and removes as objects
+  .asObjects<{name: string, GPA: RawValue}>(({adds, removes}) => {
+    for(let key in adds) {
```
Comment (joshuafcole) [inline guides/dsl.md]: I'd rename key to id for clarity -- this key will be the id of the exported record. In many cases you don't care, but in some that's useful to know.

File: guides/dsl.md [hunk 39]
```diff
@@ -0,0 +1,316 @@
...
+      // ... do something ...
```
Comment (joshuafcole) [inline guides/dsl.md]: You should show that the shape of the diff matches the shape of the record created by the watch block here.
E.g.
```
let {name, GPA} = adds[id];
// ... do something ...
```

**Review by joshuafcole — APPROVED** [review]
Rock and roll 🎸  :woman_singer:
